examples/common/hf_fsdp.py

```diff
@@ -0,0 +1,220 @@
+    # This is a hurdle for FSDP because they need to be in the same FSDP block
+    # These lines ensures that both modules stay together in the top-most block when
+    # the model has this tying enabled (almost all do; this property defaults to True)
+    if model.config.tie_word_embeddings:
+        causal_base_model._fsdp_wrap = False  # type: ignore
+        tied_embeddings._fsdp_wrap = False  # type: ignore
+        lm_head._fsdp_wrap = False  # type: ignore
+
+    # FSDP Wrap and Activation Checkpoint every model block
+    model.fsdp_wrap_fn = lambda module: isinstance(module, block_type)
+    model.activation_checkpointing_fn = lambda module: isinstance(
+        module, block_type)
+
+
+def prepare_hf_enc_dec_model_for_fsdp(model: PreTrainedModel) -> None:
+    """Wrap an encoder/decoder HF model
+
+    This works for T5, BART, Pegasus, PegasusX, but not all enc/dec (ProphetNet)
+    You have model.shared, model.encoder, model.decoder and model.lm_head,
+    where model.shared are the embeddings which are tied to model.lm_head, and
+    model.shared == model.encoder.embed_tokens and
+    model.shared == model.decoder.embed_tokens
+    """
+    tied_embeddings = model.get_input_embeddings()
+    encoder = model.get_encoder()
+    decoder = model.get_decoder()
+    lm_head = hf_get_lm_head(model)
+    # some encoder/decoders have different layers for encoder vs decoder
+    encoder_block = hf_get_hidden_layers(encoder)
+    decoder_block = hf_get_hidden_layers(decoder)
+
+    modules = {
+        "encoder": encoder,
+        "decoder": decoder,
+        "encoder_block": encoder_block,
+        "decoder_block": decoder_block,
+        "lm_head": lm_head,
+        "tied_embeddings": tied_embeddings
+    }
+
+    for mod_name, module in modules.items():
+        if module is None:
+            raise ValueError(
+                f'Unable to FSDP-wrap this model! `{mod_name}` does not ' +
+                'follow common layer/weight naming conventions.')
+    decoder_block_type = type(decoder_block[0])  # type: ignore
+    encoder_block_type = type(encoder_block[0])  # type: ignore
+
+    if model.config.tie_word_embeddings:
+        # it is possible to train an enc/dec without tied embeddings, hence the check
+        tied_embeddings._fsdp_wrap = False  # type: ignore
+        encoder._fsdp_wrap = False  # type: ignore
+        decoder._fsdp_wrap = False  # type: ignore
+        lm_head._fsdp_wrap = False  # type: ignore
+
+    # FSDP Wrap and Activation Checkpoint every decoder block
+    model.fsdp_wrap_fn = lambda module: isinstance(module, decoder_block_type)
+    model.activation_checkpointing_fn = lambda module: isinstance(
+        module, decoder_block_type)
+
+    if encoder_block_type == decoder_block_type:
+        return
+
+    # need to wrap encoder blocks separately for ProhpetNet and Marian
+    model.fsdp_wrap_fn = lambda module: isinstance(module, encoder_block_type)
```

Comment:
Do you not want `isinstance(module, (encoder_bloc_type, decoder_block_type))`, in these two functions?

Reply:
Yeah, I should move up the equality check and do that